R/sidebar.R

```diff
@@ -135,43 +130,167 @@ sidebar <- function(
+  if (!inherits(x$open, "bslib_sidebar_open_options")) {
+    # Users shouldn't arrive here, just internal code. If we're here, it's an
+    # internal error we need to fix: the render method shouldn't make changes
+    # that won't be available to caller, i.e. we should have already decided
+    # the initial sidebar `open` state.
+    rlang::abort("`sidebar$open` must be the result of `sidebar_open_on()`.")
+  }
+
+  is_always_open <- all(vapply(x$open, identical, logical(1), "always"))
+
+  if (!is_always_open && is.null(x$id)) {
+    # always provide an id for accessibility reasons when collapsible
+    x$id <- paste0("bslib-sidebar-", p_randomInt(1000, 10000))
+  }
+
+  if (!is.null(x$max_height_mobile) && x$open$mobile != "always") {
+    rlang::warn(
+      'The `max_height_mobile` argument only applies to the sidebar when `open = "always"` on mobile.'
+    )
+    x$max_height_mobile <- NULL
+  }
+
+  is_starts_closed <- all(vapply(x$open, identical, logical(1), "closed"))
+  is_starts_open <- all(vapply(x$open, identical, logical(1), "open"))
+
+  hidden_initially <-
+    if (is_always_open) FALSE
+    else if (is_starts_open) FALSE
+    else TRUE
+
   collapse_tag <-
-    if (open != "always") {
-      tags$button(
-        class = "collapse-toggle",
-        type = "button",
-        title = "Toggle sidebar",
-        "aria-expanded" = if (open %in% c("open", "desktop")) "true" else "false",
-        "aria-controls" = id,
-        collapse_icon()
-      )
+    tags$button(
+      class = "collapse-toggle",
+      type = "button",
+      title = "Toggle sidebar",
+      "aria-expanded" = tolower(is_starts_open || identical(x$open$desktop, "open")),
+      "aria-controls" = x$id,
+      collapse_icon()
+    )
+
+
+  sidebar_tag <- tags$aside(
+    id = x$id,
+    class = c("sidebar", x$class),
+    hidden = if (hidden_initially) NA,
+    tags$div(
+      class = "sidebar-content bslib-gap-spacing",
+      x$title,
+      style = css(
+        gap = x$gap,
+        padding = x$padding
+      ),
+      !!!x$attributes,
+      !!!x$children
+    )
+  )
+
+  htmltools::tagList(sidebar_tag, collapse_tag)
+}
+
+as_sidebar_open_on <- function(open) {
+  if (is.null(open)) return(NULL)
+  if (rlang::is_na(open)) open <- "always"
+  if (isTRUE(open)) open <- "open"
+  if (isFALSE(open)) open <- "closed"
+
+  if (rlang::is_character(open)) {
+    open <- sidebar_open_as_string(open, extra = "desktop", rlang::caller_env())
+    if (identical(open, "desktop")) {
+      return(sidebar_open_on("open", "always"))
     }
+    return(sidebar_open_on(open, open))
+  }
 
-  res <- list2(
-    tag = tags$aside(
-      id = id,
-      class = c("sidebar", class),
-      hidden = if (open %in% c("closed", "desktop")) NA,
-      tags$div(
-        class = "sidebar-content bslib-gap-spacing",
-        title,
-        style = css(
-          gap = validateCssUnit(gap),
-          padding = validateCssPadding(padding)
+  if (rlang::is_list(open)) {
```

Comment:
Handle this case early, then (maybe?) check length > 1, then use `sidebar_open_as_string()`?